src/data/actions/assets.js

```diff
@@ -1,74 +1,184 @@
+  (dispatch, getState) => {
+    const currentFilterState = getState().metadata.filters.assetTypes;
+
+    // because filter state is not binary, we have to add on to the request and
+    // not rely on the page metadata for constructing the request
+    const currentFilterParameters = deepCopy(getState().metadata.request.assetTypes);
+    currentFilterParameters[filterKey] = filterValue;
+
+    const parameters = {
+      assetTypes: currentFilterParameters,
+      page: 0,
+    };
+
+    return dispatch(getAssets(parameters, courseDetails)).then((responseAction) => {
+      if (requestFailed(responseAction)) {
+        dispatch(filterUpdateFailure(currentFilterState));
+      }
+    });
+  };
+
+export const clearFiltersFailure = previousFilterState => ({
+  type: assetActions.clear.CLEAR_FILTERS_FAILURE,
+  previousState: {
+    assetTypes: previousFilterState,
+  },
 });
 
-export const clearFilters = () => ({
-  type: assetActions.clear.CLEAR_FILTERS,
+export const clearFilters = courseDetails =>
+  (dispatch, getState) => {
+    const currentFilterState = getState().metadata.filters.assetTypes;
+
+    const defaultFilterParameters = getDefaultFilterState();
+    const parameters = {
+      assetTypes: defaultFilterParameters,
+      page: 0,
+    };
+
+    return dispatch(getAssets(parameters, courseDetails)).then((responseAction) => {
+      if (requestFailed(responseAction)) {
+        dispatch(clearFiltersFailure(currentFilterState));
+      }
+    });
+  };
+
+export const sortUpdateFailure = previousSortState => ({
+  type: assetActions.sort.SORT_UPDATE_FAILURE,
+  previousState: { ...previousSortState },
 });
 
-export const sortUpdate = (sort, direction) => ({
-  type: assetActions.sort.SORT_UPDATE,
-  data: { sort, direction },
+export const sortUpdate = (sort, direction, courseDetails) =>
+  (dispatch, getState) => {
+    const currentSortState = getState().metadata.sort;
+
+    const parameters = {
+      sort,
+      direction,
+    };
+
+    return dispatch(getAssets(parameters, courseDetails)).then((responseAction) => {
+      if (requestFailed(responseAction)) {
+        dispatch(sortUpdateFailure(currentSortState));
+      }
+    });
+  };
+
+export const pageUpdateFailure = previousPageState => ({
+  type: assetActions.paginate.PAGE_UPDATE_FAILURE,
+  previousState: { ...previousPageState },
 });
 
-export const pageUpdate = page => ({
-  type: assetActions.paginate.PAGE_UPDATE,
-  data: { page },
+export const resetPageState = () => ({
+  type: assetActions.paginate.RESET_PAGE,
 });
 
+export const pageUpdate = (page, courseDetails) =>
+  (dispatch, getState) => {
+    const currentPageState = getState().metadata.pagination;
+
+    const parameters = {
+      page,
+    };
+
+    return dispatch(getAssets(parameters, courseDetails)).then((responseAction) => {
+      if (requestFailed(responseAction)) {
+        // react-paginate does force the page state via the forcePage prop when
```

Comment:
Let me know if this is unclear.

Reply:
Minor clarification: this comment should read `react-paginate does` **not** `force the page state`, correct?

Reply:
Oops! Yes! Thank you for the catch.